File: projects/engineering.purdue.edu/scriptingwithobjects/swocode/chap12/CircularRef2.py
#!/usr/local/bin/python

# CircularRef2.py

import weakref                                                      #(A)

#---------------------------  class X  ---------------------------------
class X:                                                           #(X1)
    def __init__( self, id ):                                      #(X2)
        print "Creating X object: id = " + str(id)                 #(X3)
        self.id = id                                               #(X4)

    def setYfield( self, yobj ):                                   #(X5)
        self.yobj = yobj                                           #(X6)

    def __del__( self ):                                           #(X7)
        print "X obj will be destroyed: id = " + str(self.id)      #(X8)

#---------------------------  class Y  ---------------------------------
class Y:                                                           #(Y1)
    def __init__( self, id ):                                      #(Y2)
        print "Creating Y object: id = " + str(id)                 #(Y3)
        self.id = id                                               #(Y4)

    def setXfield( self, xobj ):                                   #(Y5)
        self.xobj = xobj                                           #(Y6)

    def __del__( self ):                                           #(Y7)
        print "Y obj will be destroyed: id = " + str(self.id)      #(Y8)

#--------------------------  Test Code  --------------------------------
i = 0;                                                             #(T1)
while (i < 3):                                                     #(T2)
#while (i < 100000):                                               #(T3)
    x = X(i)                                                       #(T4)
    y = Y(i)                                                       #(T5)
    x.setYfield( y )                                               #(T6)
    y.setXfield( x )                                               #(T7)
#    y.setXfield( weakref.ref( x ) )                               #(T8)
    i = i+1                                                        #(T9)

print "\n\nDONE WITH THE WHILE LOOP\n\n";                         #(T10)

File: projects/demos/engineering.purdue.edu/scriptingwithobjects/swocode/chap12/CircularRef2.pl
#!/usr/bin/perl -w

### CircularRef2.pl

use strict;

use Scalar::Util qw( weaken );

#---------------------------  class X  ---------------------------------
package X;                                                         #(X1)
    sub new {                                                      #(X2)
        my ($class, $id) = @_;                                     #(X3)
        print "Creating X object: id = " . $id . "\n";             #(X4)
        return bless { id => $id }, $class;                        #(X5)
    }
    sub setYfield {                                                #(X6)
        my ($ref, $yobj) = @_;                                     #(X7)
        $ref->{ yobj } = $yobj;                                    #(X8)
    }
    sub DESTROY {                                                  #(X9)
        my $ref = shift;                                          #(X10)
        print "X obj will be destroyed: id = " . $ref->{id} . "\n";
    }                                                             #(X11)

#---------------------------  class Y  ---------------------------------
package Y;                                                         #(Y1)
    sub new {                                                      #(Y2)
        my ($class, $id, $xobj) = @_;                              #(Y3)
        print "Creating Y object: id = " . $id . "\n";             #(Y4)
        return bless { id => $id, xob => $xobj }, $class;          #(Y5)
    }
    sub setXfield {                                                #(Y6)
        my ($ref, $xobj) = @_;                                     #(Y7)
        $ref->{ xobj } = $xobj;                                    #(Y8)
    }
    sub DESTROY {                                                  #(Y9)
        my $ref = shift;                                          #(Y10)
        print "Y obj will be destroyed: id = " . $ref->{id} . "\n";
    }                                                             #(Y11)

#--------------------------  Test Code  --------------------------------
package main;

my $i = 0;                                                         #(T1)
#while ($i < 3) {                                                  #(T2)
while ($i < 100000) {                                              #(T3)
    my $x = X->new($i);                                            #(T4)
    my $y = Y->new($i);                                            #(T5)
    $x->setYfield( $y );                                           #(T6)
    $y->setXfield( $x );                                           #(T7)
    weaken( $x->{yobj} );                                          #(T8)
    $i++;                                                          #(T9)
}                                                                 #(T10)

print "\n\nDONE WITH THE WHILE LOOP\n\n";                         #(T11)
